esp-hal/src/sha.rs

```diff
@@ -189,102 +124,105 @@ impl<'d> crate::InterruptConfigurable for Sha<'d, crate::Blocking> {
 
 // This implementation might fail after u32::MAX/8 bytes, to increase please see
 // ::finish() length/self.cursor usage
-impl<'d, DM: crate::Mode> Sha<'d, DM> {
-    pub fn first_run(&self) -> bool {
-        self.first_run
-    }
+pub trait Sha<DM: crate::Mode>: core::ops::DerefMut<Target = Context<DM>> {
+    /// Constant containing the name of the algorithm as a string.
+    const ALGORITHM: &'static str;
 
-    pub fn finished(&self) -> bool {
-        self.finished
-    }
+    /// Setup SHA Mode
+    #[cfg(not(esp32))]
+    fn mode_as_bits() -> u8;
+
+    fn chunk_length(&self) -> usize;
+
+    fn digest_length(&self) -> usize;
```

Comment:
Why wasn't this made into a constant?

Reply:
No specific reason. That might be an oversight.

Also, the trait should ideally be sealed, to not allow the user to produce wrong results by calling `write_data()`, `flush_data()`, etc. manually. But I wanted to keep it simple and as close to the original API as possible in this PR.